### ISSUE TITLE
userns-remap && overlay2  remove /usr/share/doc/ files Operation not permitted

### ISSUE THREAD

**Issue by zhangyongsu** [docker/docker#27619]
**Description**
I started a dockerd as:
/usr/bin/dockerd -H fd:// --userns-remap test --storage-driver=overlay2
when I run debian:latest image , can not remove those files in /usr/share/doc/ :

```
root@72c225848db7:/usr/share/doc# rm * -rf
rm: cannot remove 'acl': Operation not permitted
rm: cannot remove 'adduser/examples/adduser.local.conf.examples/skel.other': Operation not permitted
rm: cannot remove 'adduser/examples/adduser.local.conf.examples/skel': Operation not permitted
rm: cannot remove 'apt/examples': Operation not permitted
rm: cannot remove 'base-files': Operation not permitted
rm: cannot remove 'base-passwd': Operation not permitted
rm: cannot remove 'bash': Operation not permitted
rm: cannot remove 'bsdutils': Operation not permitted
rm: cannot remove 'coreutils': Operation not permitted
rm: cannot remove 'dash': Operation not permitted
rm: cannot remove 'debconf': Operation not permitted
rm: cannot remove 'debian-archive-keyring': Operation not permitted
rm: cannot remove 'debianutils': Operation not permitted
rm: cannot remove 'diffutils': Operation not permitted
rm: cannot remove 'dmsetup': Operation not permitted
rm: cannot remove 'dpkg': Operation not permitted
rm: cannot remove 'e2fslibs': Operation not permitted
rm: cannot remove 'e2fsprogs': Operation not permitted
rm: cannot remove 'findutils': Operation not permitted
rm: cannot remove 'gcc-4.8-base': Operation not permitted
rm: cannot remove 'gcc-4.9-base': Operation not permitted
rm: cannot remove 'gnupg/examples': Operation not permitted
rm: cannot remove 'gpgv': Operation not permitted
rm: cannot remove 'grep': Operation not permitted
rm: cannot remove 'gzip': Operation not permitted
rm: cannot remove 'hostname': Operation not permitted
rm: cannot remove 'inetutils-ping': Operation not permitted
rm: cannot remove 'init': Operation not permitted
rm: cannot remove 'initscripts': Operation not permitted
rm: cannot remove 'insserv': Operation not permitted
rm: cannot remove 'iproute2': Operation not permitted
rm: cannot remove 'libacl1': Operation not permitted
rm: cannot remove 'libapt-pkg4.12': Operation not permitted
rm: cannot remove 'libattr1': Operation not permitted
rm: cannot remove 'libaudit-common': Operation not permitted
rm: cannot remove 'libaudit1': Operation not permitted
rm: cannot remove 'libblkid1': Operation not permitted
rm: cannot remove 'libbz2-1.0': Operation not permitted
rm: cannot remove 'libc-bin': Operation not permitted
rm: cannot remove 'libc6': Operation not permitted
rm: cannot remove 'libcap2': Operation not permitted
rm: cannot remove 'libcap2-bin': Operation not permitted
rm: cannot remove 'libcomerr2': Operation not permitted
rm: cannot remove 'libcryptsetup4': Operation not permitted
rm: cannot remove 'libdb5.3': Operation not permitted
rm: cannot remove 'libdebconfclient0': Operation not permitted
rm: cannot remove 'libdevmapper1.02.1': Operation not permitted
rm: cannot remove 'libgcrypt20': Operation not permitted
rm: cannot remove 'libgpg-error0': Operation not permitted
rm: cannot remove 'libkmod2': Operation not permitted
rm: cannot remove 'liblocale-gettext-perl': Operation not permitted
rm: cannot remove 'liblzma5': Operation not permitted
rm: cannot remove 'libmount1': Operation not permitted
rm: cannot remove 'libpam-modules/examples': Operation not permitted
rm: cannot remove 'libpam-modules-bin': Operation not permitted
rm: cannot remove 'libpam-runtime': Operation not permitted
rm: cannot remove 'libpam0g': Operation not permitted
rm: cannot remove 'libpcre3': Operation not permitted
rm: cannot remove 'libprocps3': Operation not permitted
rm: cannot remove 'libreadline6/examples': Operation not permitted
rm: cannot remove 'libselinux1': Operation not permitted
rm: cannot remove 'libsemanage-common': Operation not permitted
rm: cannot remove 'libsemanage1': Operation not permitted
rm: cannot remove 'libsepol1': Operation not permitted
rm: cannot remove 'libslang2': Operation not permitted
rm: cannot remove 'libsmartcols1': Operation not permitted
rm: cannot remove 'libss2': Operation not permitted
rm: cannot remove 'libsystemd0': Operation not permitted
rm: cannot remove 'libtext-charwidth-perl': Operation not permitted
rm: cannot remove 'libtext-iconv-perl': Operation not permitted
rm: cannot remove 'libtext-wrapi18n-perl': Operation not permitted
rm: cannot remove 'libtinfo5': Operation not permitted
rm: cannot remove 'libudev1': Operation not permitted
rm: cannot remove 'libusb-0.1-4': Operation not permitted
rm: cannot remove 'libustr-1.0-1': Operation not permitted
rm: cannot remove 'libuuid1': Operation not permitted
rm: cannot remove 'login': Operation not permitted
rm: cannot remove 'lsb-base': Operation not permitted
rm: cannot remove 'mawk/examples': Operation not permitted
rm: cannot remove 'mount/examples': Operation not permitted
rm: cannot remove 'multiarch-support': Operation not permitted
rm: cannot remove 'ncurses-base': Operation not permitted
rm: cannot remove 'ncurses-bin': Operation not permitted
rm: cannot remove 'netbase': Operation not permitted
rm: cannot remove 'passwd/examples': Operation not permitted
rm: cannot remove 'perl': Operation not permitted
rm: cannot remove 'procps/examples': Operation not permitted
rm: cannot remove 'readline-common': Operation not permitted
rm: cannot remove 'sed/examples': Operation not permitted
rm: cannot remove 'sensible-utils': Operation not permitted
rm: cannot remove 'startpar': Operation not permitted
rm: cannot remove 'systemd': Operation not permitted
rm: cannot remove 'systemd-sysv': Operation not permitted
rm: cannot remove 'sysv-rc': Operation not permitted
rm: cannot remove 'sysvinit-utils': Operation not permitted
rm: cannot remove 'tar': Operation not permitted
rm: cannot remove 'tzdata': Operation not permitted
rm: cannot remove 'util-linux/releases': Operation not permitted
rm: cannot remove 'util-linux/examples': Operation not permitted
rm: cannot remove 'zlib1g': Operation not permitted

root@72c225848db7:/usr/share/doc# ls -l
total 392
drwxr-xr-x 1 root root 4096 Oct 21 03:57 acl
drwxr-xr-x 1 root root 4096 Oct 21 03:57 adduser
drwxr-xr-x 1 root root 4096 Oct 21 03:57 apt
drwxr-xr-x 1 root root 4096 Oct 21 03:57 base-files
drwxr-xr-x 1 root root 4096 Oct 21 03:57 base-passwd
drwxr-xr-x 1 root root 4096 Oct 21 03:57 bash
drwxr-xr-x 1 root root 4096 Oct 21 03:57 bsdutils
drwxr-xr-x 1 root root 4096 Oct 21 03:57 coreutils
drwxr-xr-x 1 root root 4096 Oct 21 03:57 dash
drwxr-xr-x 1 root root 4096 Oct 21 03:57 debconf
drwxr-xr-x 1 root root 4096 Oct 21 03:57 debian-archive-keyring
drwxr-xr-x 1 root root 4096 Oct 21 03:57 debianutils
drwxr-xr-x 1 root root 4096 Oct 21 03:57 diffutils
drwxr-xr-x 1 root root 4096 Oct 21 03:57 dmsetup
drwxr-xr-x 1 root root 4096 Oct 21 03:57 dpkg
drwxr-xr-x 1 root root 4096 Oct 21 03:57 e2fslibs
drwxr-xr-x 1 root root 4096 Oct 21 03:57 e2fsprogs
drwxr-xr-x 1 root root 4096 Oct 21 03:57 findutils
drwxr-xr-x 1 root root 4096 Oct 21 03:57 gcc-4.8-base
drwxr-xr-x 1 root root 4096 Oct 21 03:57 gcc-4.9-base
drwxr-xr-x 1 root root 4096 Oct 21 03:57 gnupg
drwxr-xr-x 1 root root 4096 Oct 21 03:57 gpgv
drwxr-xr-x 1 root root 4096 Oct 21 03:57 grep
drwxr-xr-x 1 root root 4096 Oct 21 03:57 gzip
drwxr-xr-x 1 root root 4096 Oct 21 03:57 hostname
drwxr-xr-x 1 root root 4096 Oct 21 03:57 inetutils-ping
drwxr-xr-x 1 root root 4096 Oct 21 03:57 init
drwxr-xr-x 1 root root 4096 Oct 21 03:57 initscripts
drwxr-xr-x 1 root root 4096 Oct 21 03:57 insserv
drwxr-xr-x 1 root root 4096 Oct 21 03:57 iproute2
drwxr-xr-x 1 root root 4096 Oct 21 03:57 libacl1
drwxr-xr-x 1 root root 4096 Oct 21 03:57 libapt-pkg4.12
drwxr-xr-x 1 root root 4096 Oct 21 03:57 libattr1
drwxr-xr-x 1 root root 4096 Oct 21 03:57 libaudit-common
drwxr-xr-x 1 root root 4096 Oct 21 03:57 libaudit1
drwxr-xr-x 1 root root 4096 Oct 21 03:57 libblkid1
drwxr-xr-x 1 root root 4096 Oct 21 03:57 libbz2-1.0
drwxr-xr-x 1 root root 4096 Oct 21 03:57 libc-bin
drwxr-xr-x 1 root root 4096 Oct 21 03:57 libc6
drwxr-xr-x 1 root root 4096 Oct 21 03:57 libcap2
drwxr-xr-x 1 root root 4096 Oct 21 03:57 libcap2-bin
drwxr-xr-x 1 root root 4096 Oct 21 03:57 libcomerr2
drwxr-xr-x 1 root root 4096 Oct 21 03:57 libcryptsetup4
drwxr-xr-x 1 root root 4096 Oct 21 03:57 libdb5.3
drwxr-xr-x 1 root root 4096 Oct 21 03:57 libdebconfclient0
drwxr-xr-x 1 root root 4096 Oct 21 03:57 libdevmapper1.02.1
drwxr-xr-x 1 root root 4096 Oct 21 03:57 libgcrypt20
drwxr-xr-x 1 root root 4096 Oct 21 03:57 libgpg-error0
drwxr-xr-x 1 root root 4096 Oct 21 03:57 libkmod2
drwxr-xr-x 1 root root 4096 Oct 21 03:57 liblocale-gettext-perl
drwxr-xr-x 1 root root 4096 Oct 21 03:57 liblzma5
drwxr-xr-x 1 root root 4096 Oct 21 03:57 libmount1
drwxr-xr-x 1 root root 4096 Oct 21 03:57 libpam-modules
drwxr-xr-x 1 root root 4096 Oct 21 03:57 libpam-modules-bin
drwxr-xr-x 1 root root 4096 Oct 21 03:57 libpam-runtime
drwxr-xr-x 1 root root 4096 Oct 21 03:57 libpam0g
drwxr-xr-x 1 root root 4096 Oct 21 03:57 libpcre3
drwxr-xr-x 1 root root 4096 Oct 21 03:57 libprocps3
drwxr-xr-x 1 root root 4096 Oct 21 03:57 libreadline6
drwxr-xr-x 1 root root 4096 Oct 21 03:57 libselinux1
drwxr-xr-x 1 root root 4096 Oct 21 03:57 libsemanage-common
drwxr-xr-x 1 root root 4096 Oct 21 03:57 libsemanage1
drwxr-xr-x 1 root root 4096 Oct 21 03:57 libsepol1
drwxr-xr-x 1 root root 4096 Oct 21 03:57 libslang2
drwxr-xr-x 1 root root 4096 Oct 21 03:57 libsmartcols1
drwxr-xr-x 1 root root 4096 Oct 21 03:57 libss2
drwxr-xr-x 1 root root 4096 Oct 21 03:57 libsystemd0
drwxr-xr-x 1 root root 4096 Oct 21 03:57 libtext-charwidth-perl
drwxr-xr-x 1 root root 4096 Oct 21 03:57 libtext-iconv-perl
drwxr-xr-x 1 root root 4096 Oct 21 03:57 libtext-wrapi18n-perl
drwxr-xr-x 1 root root 4096 Oct 21 03:57 libtinfo5
drwxr-xr-x 1 root root 4096 Oct 21 03:57 libudev1
drwxr-xr-x 1 root root 4096 Oct 21 03:57 libusb-0.1-4
drwxr-xr-x 1 root root 4096 Oct 21 03:57 libustr-1.0-1
drwxr-xr-x 1 root root 4096 Oct 21 03:57 libuuid1
drwxr-xr-x 1 root root 4096 Oct 21 03:57 login
drwxr-xr-x 1 root root 4096 Oct 21 03:57 lsb-base
drwxr-xr-x 1 root root 4096 Oct 21 03:57 mawk
drwxr-xr-x 1 root root 4096 Oct 21 03:57 mount
drwxr-xr-x 1 root root 4096 Oct 21 03:57 multiarch-support
drwxr-xr-x 1 root root 4096 Oct 21 03:57 ncurses-base
drwxr-xr-x 1 root root 4096 Oct 21 03:57 ncurses-bin
drwxr-xr-x 1 root root 4096 Oct 21 03:57 netbase
drwxr-xr-x 1 root root 4096 Oct 21 03:57 passwd
drwxr-xr-x 1 root root 4096 Oct 21 03:57 perl
drwxr-xr-x 1 root root 4096 Oct 21 03:57 procps
drwxr-xr-x 1 root root 4096 Oct 21 03:57 readline-common
drwxr-xr-x 1 root root 4096 Oct 21 03:57 sed
drwxr-xr-x 1 root root 4096 Oct 21 03:57 sensible-utils
drwxr-xr-x 1 root root 4096 Oct 21 03:57 startpar
drwxr-xr-x 1 root root 4096 Oct 21 03:57 systemd
drwxr-xr-x 1 root root 4096 Oct 21 03:57 systemd-sysv
drwxr-xr-x 1 root root 4096 Oct 21 03:57 sysv-rc
drwxr-xr-x 1 root root 4096 Oct 21 03:57 sysvinit-utils
drwxr-xr-x 1 root root 4096 Oct 21 03:57 tar
drwxr-xr-x 1 root root 4096 Oct 21 03:57 tzdata
drwxr-xr-x 1 root root 4096 Oct 21 03:57 util-linux
drwxr-xr-x 1 root root 4096 Oct 21 03:30 zlib1g
```

files seem can be remove by root in container. 

I test without --userns-remap , remove /usr/share/doc files are ok, 
without --storage-driver=overlay2 , remove /usr/share/doc files are ok, too.

**Steps to reproduce the issue:**
1.start dockerd as: /usr/bin/dockerd -H fd:// --userns-remap test --storage-driver=overlay2
2. docker run -itd debian:latest
3. docker exec -it <container id> bash
4. rm /usr/share/doc/\* -rf 
    Operation not permitted

**Output of `docker version`:**

```
Client:
 Version:      1.12.2
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   bb80604
 Built:        Tue Oct 11 17:43:41 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.2
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   bb80604
 Built:        Tue Oct 11 17:43:41 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 1
 Running: 1
 Paused: 0
 Stopped: 0
Images: 1
Server Version: 1.12.2
Storage Driver: overlay2
 Backing Filesystem: extfs
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: null overlay host bridge
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options:
Kernel Version: 4.4.0-0.bpo.1-amd64
Operating System: Debian GNU/Linux 8 (jessie)
OSType: linux
Architecture: x86_64
CPUs: 40
Total Memory: 62.56 GiB
Name: n3-021-211
ID: O3MK:N5ER:PKZB:IKNW:U5ZF:LA6Q:PNYG:LJ6T:VX7Z:D7I4:TUXB:LW6F
Docker Root Dir: /var/lib/docker/624288.624288
Debug Mode (client): false
Debug Mode (server): false
Http Proxy: http://10.4.16.31:3128/
No Proxy: localhost,127.0.0.1,dockerhub.byted.org
Registry: https://index.docker.io/v1/
WARNING: No kernel memory limit support
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

**Comment by justincormack** [comment]
Simpler repro is just : `docker run debian rm -rf /usr/share/doc`

**Comment by justincormack** [comment]
cc @estesp

**Comment by justincormack** [comment]
Does not happen on `aufs`.

**Comment by justincormack** [comment]
cc @dmcgowan as overlay related

**Comment by estesp** [comment]
This is kernel dependent..maybe @runcom knows more about when the overlay + userns fixes went upstream?

I'm on `Linux ubu1604vm 4.4.0-45-generic` and the `rm -fr /usr/share/doc` works properly on **overlay** and **overlay2**

**Comment by shadeslayer** [comment]
I can definitely reproduce this too.

**Comment by dmcgowan** [comment]
@shadeslayer could you provide your `docker info`

**Comment by shadeslayer** [comment]
```
Containers: 6
 Running: 1
 Paused: 0
 Stopped: 5
Images: 4
Server Version: 1.12.1
Storage Driver: overlay
 Backing Filesystem: extfs
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: host bridge null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options:
Kernel Version: 4.5.7-std-4
Operating System: Ubuntu 16.04 LTS
OSType: linux
Architecture: armv7l
CPUs: 4
Total Memory: 1.974 GiB
Name: alpha-arm-16-04
ID: TY2T:BIEN:AS6Z:BLF2:DGFP:3GAM:VYZI:M2IN:Q4OF:ZMLY:JH7C:R73L
Docker Root Dir: /var/lib/docker/100000.120
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No cpu cfs quota support
WARNING: No cpu cfs period support
Insecure Registries:
 127.0.0.0/8
```

We're on a slightly exotic architecture but running docker from the ubuntu xenial packages.

**Comment by shadeslayer** [comment]
I've even tried to switch to the devicemapper backend, but unfortunately docker just hangs after it creates the pools there :(

**Comment by dmcgowan** [comment]
@shadeslayer we are still trying to pin down which kernel versions are effected, if it is a more recent fix then the maintained 4.4 kernel is more likely to have the fix if that is something you are interested in trying out.

**Comment by zhangyongsu** [comment]
hi @estesp because I use debian. the kernel 4.4 can not update by debian now . so I compiled a 4.4.33 (almost the newest version) for my debian . and this issue can not be reproduced. 
so seems not the docker's problem. 
I close it .and update kernel version to solution. thanks all.
